### PR TITLE
[BULLET] Making GetContactsFromLastStepFeature optional in Collision Features

### PR DESCRIPTION
**Pull request by Lobotuerk**
# 🎉 New feature

Makes GetContactFromLastStepFeture not a requirement for Collisions in IGN-Physics. Fixes https://github.com/ignitionrobotics/ign-physics/pull/208#discussion_r594540971

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Comment by azeey** [comment]
This is not compiling mainly because of https://github.com/ignitionrobotics/ign-gazebo/blob/f5a362cf579ef23d2cd9d5ec1cd2f856410db3c3/src/systems/physics/Physics.cc#L2202-L2203

The `EntityFeatureMap::Get` expects an `EntityPtr` with required/minimum FeatureList. In this snippet of code `contact.collision1` has `ContactFeatureList`, which is not the required feature in `EntityCollisionMap`. Ideally, this should downcast cast to the entity with the required FeatureList, but there's an implicit conversion issue that causes an ambiguity in the compiler: the compiler is unable to decide whether to call `EntityFeatureMap::Get(const Entity &)` or `EntityFeatureMap::Get(const RequiredEntityPtr &_physEntity)`. The temporary solution is to manually create an ShapePtr with the required features from `contact.collision1` before calling `EntityFeatureMap::Get`. 

```diff
diff --git a/src/systems/physics/Physics.cc b/src/systems/physics/Physics.cc
index 9b419b8c..5ce52614 100644
--- a/src/systems/physics/Physics.cc
+++ b/src/systems/physics/Physics.cc
@@ -313,3 +313,3 @@ class ignition::gazebo::systems::PhysicsPrivate
   public: using ShapePtrType = ignition::physics::ShapePtr<
-            ignition::physics::FeaturePolicy3d, ContactFeatureList>;
+            ignition::physics::FeaturePolicy3d, CollisionFeatureList>;
 
@@ -432,2 +432,3 @@ class ignition::gazebo::systems::PhysicsPrivate
             CollisionFeatureList,
+            ContactFeatureList,
             CollisionMaskFeatureList,
@@ -2201,5 +2202,6 @@ void PhysicsPrivate::UpdateCollisions(EntityComponentManager &_ecm)
     const auto &contact = contactComposite.Get<WorldShapeType::ContactPoint>();
-    auto coll1Entity = this->entityCollisionMap.Get(contact.collision1);
-    auto coll2Entity = this->entityCollisionMap.Get(contact.collision2);
-
+    auto coll1Entity =
+        this->entityCollisionMap.Get(ShapePtrType(contact.collision1));
+    auto coll2Entity =
+        this->entityCollisionMap.Get(ShapePtrType(contact.collision2));
```

I'll continue to think about a way to fix the implicit conversion problem, but I don't think it's needed for this PR.

File: src/systems/physics/Physics.cc [hunk 1]
```diff
@@ -373,6 +382,7 @@ class ignition::gazebo::systems::PhysicsPrivate
           physics::World,
           MinimumFeatureList,
           CollisionFeatureList,
+          ContactFeatureList,
```
Comment (Blast545) [inline src/systems/physics/Physics.cc:385]: If adding this, `CollisionFeatureList,` can be removed

Comment (Lobotuerk) [reply src/systems/physics/Physics.cc:385]: Following that logic, `MinimumFeatureList` should have been removed before, I'm leaving them all just in case

Comment (azeey) [reply src/systems/physics/Physics.cc:380]: I agree with @Lobotuerk, I think we want to support physics engines that support `CollisionFeatureList`, but not `ContactFeatureList`.

**Review by azeey — APPROVED** [review]
The failing test on Ubuntu is also failing on our ign-gazebo4 CI https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo4-bionic-amd64/54/, so it's not caused by this PR.

**Review by Blast545 — APPROVED** [review]
LGTM

**Review by gonzodepedro — APPROVED** [review]
LGTM

**Pull request merged**